src/Native/Unix/System.Native/pal_uid.c

```diff
@@ -86,7 +98,73 @@ uint32_t SystemNative_GetEGid()
     return getegid();
 }
 
-int32_t SystemNative_SetEUid(uid_t euid)
+int32_t SystemNative_SetEUid(uint32_t euid)
 {
     return seteuid(euid);
 }
+
+uint32_t SystemNative_GetUid()
+{
+    return getuid();
+}
+
+#ifdef USE_GROUPLIST_LOCK
+static pthread_mutex_t s_groupLock = PTHREAD_MUTEX_INITIALIZER;
+#endif
+
+int32_t SystemNative_GetGroupList(const char* name, uint32_t group, uint32_t* groups, int32_t* ngroups)
+{
+    assert(name != NULL);
+    assert(groups != NULL);
+    assert(ngroups != NULL);
+    assert(*ngroups >= 0);
+
+    // The man page of getgrouplist doesn't explicitly mention the use of errno on error.
+    // glibc and musl implementation look like they return -1 for functions that fail with errno set.
+    // To be safe, we'll assume errno is set on error and handle EINTR.
+    int rv;
+    int groupsAvailable;
+    do
+    {
+        errno = 0;
+        groupsAvailable = *ngroups;
+
+#ifdef USE_GROUPLIST_LOCK
+        rv = pthread_mutex_lock(&s_groupLock);
+        if (rv != 0)
+        {
+            errno = rv;
+            rv = -1;
+            break;
+        }
+#endif
+
+#ifdef __APPLE__
+        // On OSX groups are passed as a signed int.
+        rv = getgrouplist(name, (int)group, (int*)groups, &groupsAvailable);
+#else
+        rv = getgrouplist(name, group, groups, &groupsAvailable);
+#endif
+
+#ifdef USE_GROUPLIST_LOCK
+        pthread_mutex_unlock(&s_groupLock);
```

Comment:
Is it guaranteed this won't mess with errno?

Reply:
Yes, most pthread APIs (including this one) return an error code instead of using errno.